File: trees/LCA.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* left;
        Node* right;

        Node(int data) {
            this->data = data;
            this->left = NULL;
            this->right = NULL;
        }
};

class Tree {
    public:
        Node* root;

        Tree() {
            this->root = NULL;
        }

        void constructTree() {
            Node* root = new Node(8);

            root->left = new Node(3);
            root->left->left = new Node(1);
            root->left->right = new Node(6);
            root->left->right->left = new Node(4);
            root->left->right->right = new Node(7);

            root->right = new Node(12);
            root->right->left = new Node(11);
            root->right->right = new Node(15);
            root->right->right->left = new Node(14);

            this->root = root;
            return;
        }

        void findLCA(Node* root, int node1, int node2) {
            if(root == NULL) {
                return;
            }

            bool node1_exists = false;
            bool node2_exists = false;

            Node* lca = findLCAUtil(root, node1, node1_exists, node2, node2_exists);
            if(node1_exists && node2_exists) {
                cout << "Found LCA: " << lca->data << endl;
            } else if(node1_exists && !node2_exists) {
                cout << node2 << " does not exist" << endl;
            } else if(!node1_exists && node2_exists) {
                cout << node1 << " does not exist" << endl;
            } else {
                cout << node1 << " and " << node2 << " do not exist" << endl;
            }

            return;
        }

        void destroyTree(Node* root) {
            if(root == NULL) {
                return;
            }

            destroyTree(root->left);
            destroyTree(root->right);
            root->left = NULL;
            root->right = NULL;
            delete root;
        }

    private:
        Node* findLCAUtil(Node* root, int node1, bool &node1_exists, int node2, bool &node2_exists) {

            if(root == NULL) {
                return NULL;
            }

            // This will store LCA to return the search earlier
            Node* test = NULL;

            if(root->data == node1) {
                node1_exists = true;
                test = root;
            }

            if(root->data == node2) {
                node2_exists = true;
                test = root;
            }

            // If LCA already found, return it.
            if(node1_exists && node2_exists && test != NULL) {
                return test;
            }

            Node* LCALeft = findLCAUtil(root->left, node1, node1_exists, node2, node2_exists);
            Node* LCARight = findLCAUtil(root->right, node1, node1_exists, node2, node2_exists);

            // check if temp is set, which means LCA was found
            if(test != NULL) {
                return test;
            }

            if(LCALeft == NULL && LCARight == NULL) {
                return NULL;
            } else if(LCALeft != NULL && LCARight != NULL) {
                return root;
            } else {
                return (LCALeft != NULL) ? LCALeft : LCARight;
            }
        }
};

int main() {
    Tree tree;
    tree.constructTree();

    tree.findLCA(tree.root, 14, 11);
    tree.findLCA(tree.root, 8, 3);

    tree.destroyTree(tree.root);
    tree.root = NULL;

    return 0;
}

File: trees_heaps/trees_py/LCA.py
"""
Find the lowest common ancestor of 2 tree nodes
"""

# -*- encoding: utf-8 *-*


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Tree:
    def __init__(self):
        self.root = None

    def construct_tree(self):
        self.root = Node(8)
        self.root.left = Node(3)
        self.root.left.left = Node(1)
        self.root.left.right = Node(6)
        self.root.left.right.left = Node(4)
        self.root.left.right.right = Node(7)
        self.root.right = Node(12)
        self.root.right.left = Node(11)
        self.root.right.right = Node(15)
        self.root.right.right.left = Node(14)

    def print_level_order_traversal(self, root):
        if root is None:
            return None

        queue = []
        queue.append(root)
        print()
        while(len(queue) > 0):
            count = len(queue)
            while(count > 0):
                curr_node = queue.pop(0)
                print(curr_node.data, end=" ")
                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)
                count -= 1
            print()

    def find_LCA_util(self, root, val1_exists, val1, val2_exists, val2):
        temp = Node(None)
        if root is None:
            return None

        if root.data == val1:
            val1_exists[0] = True
            temp = root

        if root.data == val2:
            val2_exists[0] = True
            temp = root

        if val1_exists[0] and val2_exists[0] and temp.data:
            return temp

        lca_left = self.find_LCA_util(root.left, val1_exists, val1, val2_exists, val2)
        lca_right = self.find_LCA_util(root.right, val1_exists, val1, val2_exists, val2)

        if temp.data:
            return temp

        if not lca_left and not lca_right:
            return None
        elif lca_left and lca_right:
            return root
        elif lca_left:
            return lca_left
        else:
            return lca_right

    def find_LCA(self, root, val1, val2):
        if root is None:
            return

        val1_exists = [False]
        val2_exists = [False]
        lca = self.find_LCA_util(root, val1_exists, val1, val2_exists, val2)
        if val1_exists[0] and val2_exists[0]:
            print("LCA - ", lca.data)
        elif not val1_exists[0] and not val2_exists[0]:
            print(val1, ", and ", val2, " does not exist")
        elif val2_exists[0]:
            print(val1, " does not exist")
        else:
            print(val2, " does not exist")


def main():
    tree = Tree()
    tree.construct_tree()
    # tree.print_level_order_traversal(tree.root)
    tree.find_LCA(tree.root, 14, 11)
    return


if __name__ == '__main__':
    main()
